### DISCUSSION TITLE
How to load custom *remote* font with Next.js and Vercel?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#56684] [Help]
### Summary

I am looking at https://nextjs.org/docs/app/api-reference/components/font and https://nextjs.org/docs/pages/building-your-application/optimizing/fonts but they only show:

```ts
import localFont from 'next/font/local'
 
const myFont = localFont({ src: './my-font.woff2' })
```

What if my font is hosted in an s3 bucket? Or some other remote URL? How can I load it then? I have 10,000 fonts, I think it is too many to load "locally" in a Vercel app, I would prefer to have this many fonts loaded in a dedicated CDN. Something like this would be great:

```ts
import localFont from 'next/font/local'
 
const myFont = localFont({ src: 'https://cdn.mysite.com/fonts/MyCustomFont-Regular.ttf' })
```

Or even:

```ts
import remoteFont from 'next/font/remote'
 
const myFont = remoteFont({ src: 'https://cdn.mysite.com/fonts/MyCustomFont-Regular.ttf' })
```

Thanks for your help!

**Comment by glomotion** [comment]
I'd also love to learn about this!

**Comment by bhaveyx** [comment]
Any update on this? Came across a similar situation.

**Comment by looterz** [comment]
Is there really no native way to use fonts you have hosted on your own CDN like Cloudflare R2? Just google and local fonts only?

**Comment by dnywh** [comment]
+1. Would love to remote fonts supported in next/font. 

Here’s how I’ve done it in the interim, for any lurkers reading:

I’m already using [Pigment CSS](https://github.com/mui/pigment-css?tab=readme-ov-file) (like styled-components). Pigment CSS supports and recommends setting a single global CSS instance in the root `app/layout.tsx` file. I just added the following to the top of that global CSS:

```tsx
import { globalCss } from "@pigment-css/react";

globalCss`
  @font-face {
    font-family: 'National';
    src: url('https://mystoragelocation.example/national-2-regular.woff2') format('woff2');
    font-weight: 500;
    font-style: normal;
    font-display: swap;
  }
  // Repeat above for the other font-weights

  // Apply that font family to all elements
  * {
    font-family: 'National', system-ui, -apple-system, sans-serif;
  }

```

It’s not the cleanest solution, and misses out on some next/font niceties (e.g. subsetting), but it works.

I subset (subsetted?) the font manually with fonttools and used that as my remote font source:

```python
# Install fonttools and brotli
pip install fonttools
pip install brotli

# Loop through all National font files
for font in national-2-*.woff2; do
    # Extract the name without extension
    basename="${font%.woff2}"
    
    echo "Subsetting $font..."
    
    pyftsubset "$font" \
        --output-file="${basename}-subset.woff2" \
        --flavor=woff2 \
        --layout-features='*' \
        --unicodes="U+0000-00FF" \
        --desubroutinize
done

echo "Done! Created subset versions of all fonts"
```

**Comment by vivshaw** [comment]
Yet another upvote. Currently it's challenging to use `next/font` with paid fonts. Many type foundries have licenses that permit web use, but forbid storing in a public Git repo. It's easy to comply with these licenses by hosting the font in some other way that does not involve sticking it in your `/public` directory, such as a CDN, but that's not compatible with `next/font`. This leaves you with the choice of either closed-sourcing your Next app or giving up on `next/font` and eating the performance hit. It would be excellent if I could instead embed it at build time from a remote origin, like `next/font` can do with Google Fonts.

**Comment by alexanderbnelson** [comment]
Yet another upvote. Seems like such a basic feature. remoteFont() would solve being able to use the entire Adobe Fonts library in one fell swoop.
